torchvision/models/convnext.py

```diff
@@ -0,0 +1,201 @@
+        s += ", out_channels={out_channels}"
+        s += ", num_layers={num_layers}"
+        s += ")"
+        return s.format(**self.__dict__)
+
+
+class ConvNeXt(nn.Module):
+    def __init__(
+        self,
+        block_setting: List[CNBlockConfig],
+        stochastic_depth_prob: float = 0.0,
+        layer_scale: float = 1e-6,
+        num_classes: int = 1000,
+        block: Optional[Callable[..., nn.Module]] = None,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__()
+        _log_api_usage_once(self)
+
+        if not block_setting:
+            raise ValueError("The block_setting should not be empty")
+        elif not (isinstance(block_setting, Sequence) and all([isinstance(s, CNBlockConfig) for s in block_setting])):
+            raise TypeError("The block_setting should be List[CNBlockConfig]")
+
+        if block is None:
+            block = CNBlock
+
+        if norm_layer is None:
+            norm_layer = partial(LayerNorm2d, eps=1e-6)
+
+        layers: List[nn.Module] = []
+
+        # Stem
+        firstconv_output_channels = block_setting[0].input_channels
+        layers.append(
+            ConvNormActivation(
+                3,
+                firstconv_output_channels,
+                kernel_size=4,
+                stride=4,
+                padding=0,
+                norm_layer=norm_layer,
+                activation_layer=None,
+                bias=True,
+            )
+        )
+
+        total_stage_blocks = sum(cnf.num_layers for cnf in block_setting)
+        stage_block_id = 0
+        for cnf in block_setting:
+            # Bottlenecks
+            stage: List[nn.Module] = []
+            for _ in range(cnf.num_layers):
+                # adjust stochastic depth probability based on the depth of the stage block
+                sd_prob = stochastic_depth_prob * stage_block_id / (total_stage_blocks - 1.0)
+                stage.append(block(cnf.input_channels, layer_scale, sd_prob, norm_layer))
+                stage_block_id += 1
+            layers.append(nn.Sequential(*stage))
+            if cnf.out_channels is not None:
+                # Downsampling
+                layers.append(
+                    nn.Sequential(
+                        norm_layer(cnf.input_channels),
+                        nn.Conv2d(cnf.input_channels, cnf.out_channels, kernel_size=2, stride=2),
+                    )
+                )
+
+        self.features = nn.Sequential(*layers)
+        self.avgpool = nn.AdaptiveAvgPool2d(1)
+
+        lastblock = block_setting[-1]
+        lastconv_output_channels = (
+            lastblock.out_channels if lastblock.out_channels is not None else lastblock.input_channels
+        )
+        self.classifier = nn.Sequential(
+            norm_layer(lastconv_output_channels), nn.Flatten(1), nn.Linear(lastconv_output_channels, num_classes)
+        )
+
+        for m in self.modules():
```

Comment:
Also, if further along some standardization of reset_parameters is envisioned (https://github.com/pytorch/vision/issues/3410), at least moving initialization to a separate function could be useful

Reply:
Thanks for the proposal. Unfortunately the approach we use for initializing the layers does not play nice with FX as the information of classes are lost and we can't use `isinstance()`. So while this check works in the constructor, after the model gets traced it won't be possible to be reset.

This is something raised previously at https://github.com/pytorch/pytorch/issues/66335 if you want to read more info. Until this use-case gets better support, we can't use both FX and making public a `reset_parameters` method.

Reply:
As for the proposal of using Sequential, I'm also a very big fan of implementing the models as Sequential (This is why the `features` of this implementation uses it). Unfortunately, there are some limitations imposed by quantization and JIT that require defining explicitly a forward and even using a `_forward_impl()` (you can see an example of the idiom [here](https://github.com/pytorch/vision/blob/main/torchvision/models/quantization/mobilenetv3.py#L98-L102)). So many of these not-so-elegant approaches are done to circumvent known limitations but I'm hopeful that we will be able to get around them soon.

Reply:
My 

> making public a `reset_parameters` method.

Well, even if it's a non-public randomly named `_reset_parameters`, `_init_parameters` etc, this will make further uniformization easier. And maybe if jitted model calls to it are not supported yet, having it named as in nn.Linear helps readability etc.

Reply:
> Unfortunately the approach we use for initializing the layers does not play nice with FX as the information of classes are lost and we can't use `isinstance()`

The fact of FX blocking subclassing in such ways is a pretty serious degradation...

Reply:
> As for the proposal of using Sequential, I'm also a very big fan of implementing the models as Sequential (This is why the `features` of this implementation uses it). Unfortunately, there are some limitations imposed by quantization and JIT that require defining explicitly a forward and even using a `_forward_impl()` (you can see an example of the idiom [here](https://github.com/pytorch/vision/blob/main/torchvision/models/quantization/mobilenetv3.py#L98-L102)). So many of these not-so-elegant approaches are done to circumvent known limitations but I'm hopeful that we will be able to get around them soon.

In that old issue, I proposed of just inserting quant/dequant stubs in the sequential itself, then there should be no problem with forward_impl, right?

Reply:
Unfortunately not everything can be implemented neatly as a Sequential ([example](https://github.com/pytorch/vision/blob/main/torchvision/models/quantization/mobilenetv3.py#L28-L29)). Though I do like to simplify the code, I wouldn't want to introduce things if they are not fully supported because it can bite downstream users. 

For example, introducing a `_reset_parameters` (which we've done previously in [RegNets](https://github.com/pytorch/vision/blob/main/torchvision/models/regnet.py#L370) before we learn about the FX limitation and we should remove) will give the ability to users to call it whenever they want, even if it's private. But then if they combine this with the FX-base feature extractor, it won't work. Keeping it in the constructor where it's inaccessible makes it a responsibility of the end user to write a method to re-init the model in cases where this is allowed.

This discussion is important and I'm glad you raised it. Let's continue it on the respective tickets you opened so that that this doesn't get buried in this unrelated PR.

Reply:
> I wouldn't want to introduce things if they are not fully supported because it can bite downstream users.

Well, currently FX feature extractor isn't fully supported wrt hooks (which is an old core feature), it hasn't blocked its introduction and publicizing. FX is a nice feature, but it's not seamless, e.g. it blocks any existing module-based transformations, as it erases moduels.

> will give the ability to users to call it whenever they want, even if it's private. B

This also has never been a problem in core pytorch, plenty of private underscore-prefixed methods exist in all sort of classes. And them being private is sufficient for not worrying people would complain of them problems. Anyway, `reset_parameters`-kind of methods already exist in almost all models.

Reply:
> Unfortunately not everything can be implemented neatly as a Sequential ([example](https://github.com/pytorch/vision/blob/main/torchvision/models/quantization/mobilenetv3.py#L28-L29)). Though I do like to simplify the code, I wouldn't want to introduce things if they are not fully supported because it can bite downstream users.

Of course, not everything is sequential (I would say, this is a straw argument), but backbones more often than not are and can be quantized just with inserting stubs without more complex forward_impl idiom, right? unless I'm missing some other problems of quantization